Bottom: e5cad329c8c5675b412e1196fe6477e3ff90a48f
Top:    cb13e87a22ec7057438542d7607ee64dc5e68834
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-04-24 00:25:12 +0100

Refresh of draw-c-change-drawtop-to-drawo

---

diff --git a/src/draw.c b/src/draw.c
index 4cac1d8..2a097d9 100644
--- a/src/draw.c
+++ b/src/draw.c
@@ -83,7 +83,7 @@ static const BoxType *clip_box = NULL;
  */
 static void Redraw (bool, BoxTypePtr);
 static void DrawEverything (BoxTypePtr);
-static void DrawPPV (int side, const BoxType *);
+static void DrawPPV (int group, const BoxType *);
 static int DrawLayerGroup (int, const BoxType *);
 static void DrawPinOrViaLowLevel (PinTypePtr, bool);
 static void DrawPlainPin (PinTypePtr, bool);
@@ -316,7 +316,7 @@ PrintAssembly (const BoxType * drawn_area, int side_group, int swap_ident)
 
   gui->set_draw_faded (Output.fgGC, 1);
   DrawLayerGroup (side_group, drawn_area);
-  DrawPPV (side, drawn_area);
+  DrawPPV (side_group, drawn_area);
   gui->set_draw_faded (Output.fgGC, 0);
 
   /* draw package */
@@ -403,11 +403,9 @@ DrawEverything (BoxTypePtr drawn_area)
   if (TEST_FLAG (CHECKPLANESFLAG, PCB) && gui->gui)
     return;
 
-  side = SWAP_IDENT ? SOLDER_LAYER : COMPONENT_LAYER;
-
   /* Draw pins, pads, vias below silk */
   if (gui->gui)
-    DrawPPV (side, drawn_area);
+    DrawPPV (SWAP_IDENT ? solder : component, drawn_area);
   else
     {
       HoleCountStruct hcs;
@@ -569,18 +567,33 @@ pad_callback (const BoxType * b, void *cl)
 }
 
 /* ---------------------------------------------------------------------------
- * draws outer layer pins pads and vias
+ * draws pins pads and vias
  */
 static void
-DrawPPV (int side, const BoxType * screen)
+DrawPPV (int group, const BoxType * screen)
 {
+  int component_group = GetLayerGroupNumberByNumber (component_silk_layer);
+  int solder_group = GetLayerGroupNumberByNumber (solder_silk_layer);
+
   if (PCB->PinOn || doing_assy)
     {
       /* draw element pins */
       r_search (PCB->Data->pin_tree, screen, NULL, pin_callback, NULL);
+
       /* draw element pads */
-      r_search (PCB->Data->pad_tree, screen, NULL, pad_callback, &side);
+      if (group == component_group)
+        {
+          side = COMPONENT_LAYER;
+          r_search (PCB->Data->pad_tree, drawn_area, NULL, pad_callback, &side);
+        }
+
+      if (group == solder_group)
+        {
+          side = COMPONENT_LAYER;
+          r_search (PCB->Data->pad_tree, drawn_area, NULL, pad_callback, &side);
+        }
     }
+
   /* draw vias */
   if (PCB->ViaOn || doing_assy)
     {
